Firestore/Source/Local/FSTLevelDB.mm

```diff
@@ -309,6 +309,15 @@ - (instancetype)initWithDirectory:(firebase::firestore::util::Path)directory
     _referenceDelegate =
         [[FSTLevelDBLRUDelegate alloc] initWithPersistence:self lruParams:lruParams];
     _transactionRunner.SetBackingPersistence(self);
+    Status status = [self start];
+    if (!status.ok()) {
+      // If leveldb fails to start then just throw up our hands: the error is unrecoverable.
+      // There's nothing an end-user can do and nearly all failures indicate the developer is doing
+      // something grossly wrong so we should stop them cold in their tracks with a failure they
+      // can't ignore.
+      [NSException raise:NSInternalInconsistencyException
```

Comment:
This doesn't really work once we transition to C++ where we won't have exceptions and there's no opportunity to fail in a constructor. Meanwhile, in C++ there is an opportunity to fail to initialize a Firebase component, so this change really ends up being a step in the wrong direction.

I think we can still remove start from the general contract of persistence, but we still need to be able to fail in a way that can be plumbed out to a user without an exception.

Reply:
I assume we would `HARD_ASSERT`, which I assume we can do in a constructor. We don't expect users to recover from this exception in the current state, right?

Reply:
Or, since two out of three persistence options currently go through a static factory method, we could port the last one to do that and have them all return `StatusOr`, if we think that's friendlier.

Reply:
We can HARD_ASSERT in a constructor.

I don't think that's the right vehicle for this particular problem though since HARD_ASSERT means there's our bug in the code. We specifically didn't do that previously in order to avoid conveying that it was an internal problem.

The C++ folks seem to want us to avoid crashing so the StatusOr thing seems like a good path.